src/ripple/overlay/MessageScheduler.md

```diff
@@ -0,0 +1,350 @@
+- **receiver**: An object that is waiting for a response.
+
+
+## Interface
+
+Let us walk through the typical workflow of a request from the perspective of
+a dependent.
+
+The first step is to schedule a **sender**, adding it to a queue of senders
+ordered by priority.[^5]
+A sender represents a demand for at least one open channel to which to send
+messages.
+Each channel represents the capacity to send one message to one peer.
+The sender has a callback method, `onOffer`, that accepts an **offer**.
+The `onOffer` callback may be called immediately,
+in the same thread that scheduled the sender.
+If the caller wants the offer to come in a different thread,
+perhaps to avoid blocking the current thread or double locking a mutex,
+then it should schedule a job that will schedule the sender in another thread.
+
+[^5]: For now, the priority order is just first come, first served.
+We want to consider more sophisticated priority schemes for the future.
+
+Eventually an offer will be made by passing it to `sender.onOffer`.
+An offer contains N open channels,
+and permits the sender to send messages to a subset M &lt;= N of them.
+M is called the **size** of the offer.
+The sender can iterate through the open channels in the offer,
+selecting the ones it likes,
+sending messages to as many or as few as it wants,
+until (a) it has no more messages to send,
+(b) the offer is exhausted, or
+(c) it has considered every channel in the offer.
+Sending a message to a channel closes the channel.
+Once a channel is closed, it may not be used to send another message.
+
+If a sender expects a response to its message, then we call that message
+a **request**.
+If it does not expect a response to its message, then we call that message
+a **notification**.
+Whenever it sends a request,
+the sender must pass to the scheduler a **receiver**
+that will wait for and react to the response.
+
+The sender must set a timeout for each message it sends.
+For requests, this timeout reflects the time after which a response is
+considered missing and the sender may retry.
+For notifications, this timeout reflects the time it expects the peer to need
+to digest the notification,
+effectively limiting the message rate of senders.
+
+If the sender closed any channels, it will be removed from the sender queue.
+If it closed no channels, perhaps because it is waiting for a specific peer,
+then it will be left in the queue to see another, different offer later.
+If it closed a channel, but wants more, perhaps because it has more messages
+to send than there were open channels in the offer,
+then it must schedule another sender.
+
+When the response arrives, the receiver's success callback will be called with
+it.
+If, before that happens, the timeout expires or the peer disconnects,
+then the receiver's failure callback will be called with an error code.
+Either way, it may choose to schedule another sender, looping back around.
+
+
+## Design
+
+### Offers
```

Comment:
The thesaurus recommends "proposal" as an alternative. :joy: 

`Channels` could be fine. I'd be fine with `ChannelSet` too. What would we call the event, i.e. how would we rename `onOffer`?

Reply:
I think both `Channels` and `ChannelSet` would work well. The event might be named `onChannelAvail` or something like that?

Reply:
How about `onReady`?

Reply:
See https://github.com/XRPLF/rippled/pull/4217#discussion_r1151011183.

Reply:
Ok, I'm going with a different design for the offers. I'm calling the object a `Courier`: senders send messages through it, and it has a limited capacity. I'm still open to suggestions for a name.

`MessageScheduler` no longer holds a `std::vector<Channel>`. Instead, it has a `std::list<MetaPeer>` (for storage) and a `std::vector<typename std::list<MetaPeer>::iterator>`, called a `MetaPeerSet` or "peer set". The peer set is sorted by ID for fast set operations.

Each `MetaPeer` tracks a number of total channels and closed channels, and the `MessageScheduler` tracks the sums of these numbers.

A `Courier` holds a reference to the `MessageScheduler`. It can return the full set of peers, including peers that have no open channels, and peers that are disconnected (their `std::weak_ptr` cannot be locked) but not yet removed. This way senders can always look at the full set of peers, not just the ones that recently opened a channel.

The `Courier` places a limit on the number of messages that a sender can send, to implement fair scheduling.

Messages must be sent through the `Courier` with a `MetaPeer`, not through a `std::shared_ptr<Peer>` directly, so that it can correctly track closed channels.

Reply:
`Courier` seems like it should work well.  Thanks.